File: WindowsCE/NurSample_vs2008_VB/NurSampleCE_VB/BeeperLocator.vb
Imports System.Collections.Generic
Imports System.Threading

Class BeeperLocator
    Implements IDisposable
    Private m_maxLevel As Integer = 100
    Private beeperRate As Integer = 0
    Private beeperStopped As Boolean = True
    Private beeperRunning As Boolean = False
    Private beeper As New Beeper()
    Private hBeeperThread As Thread = Nothing
    Private beepEvent As New ManualResetEvent(False)

    Public Sub New()
        ' Start Beeper thread
        beeperRate = 0
        beeperRunning = True
        hBeeperThread = New Thread(AddressOf BeeperThread)
        hBeeperThread.IsBackground = True
        hBeeperThread.Priority = ThreadPriority.BelowNormal
        hBeeperThread.Start()
    End Sub

    Public Sub Dispose() Implements System.IDisposable.Dispose
        ' Stop and Exit Beeper thread
        beeperRunning = False
        beepEvent.[Set]()
    End Sub

    Public Sub Beep(ByVal level As Integer, ByVal force As Boolean)
        beeperRate = (level * 100) / m_maxLevel
        If force Then
            beepEvent.[Set]()
        End If
    End Sub

    Public Property MaxLevel() As Integer
        Get
            Return m_maxLevel
        End Get
        Set(ByVal value As Integer)
            m_maxLevel = value
        End Set
    End Property

    Public Sub Start()
        beeperStopped = False
        beepEvent.[Set]()
    End Sub

    Public Sub [Stop]()
        beeperStopped = True
        beepEvent.[Set]()
    End Sub

    Private Sub BeeperThread()
        While beeperRunning
            beepEvent.Reset()
            If beeperStopped Then
                ' State: Stopped
                beepEvent.WaitOne()
            Else
                ' State: Locating
                If beeperRate > 0 Then
                    ' Tag in the range
                    beeper.Hz = 1568
                    If beeperRate >= 100 Then
                        beeper.Beep(100, False)
                    Else
                        beeper.Beep(50, False)
                        beepEvent.WaitOne(250 - beeperRate, False)
                    End If
                Else
                    ' No tag in the range
                    beeper.Hz = 1046
                    beeper.Beep(50, False)
                    beepEvent.WaitOne(2000, False)
                End If
            End If
        End While
        hBeeperThread = Nothing
    End Sub
End Class

File: WindowsCE/NurSample_vs2008/NurSample/BeeperLocator.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace NurSample
{
    class BeeperLocator : IDisposable
    {
        int maxLevel = 100;
        int beeperRate = 0;
        bool beeperStopped = true;
        bool beeperRunning = false;
        Beeper beeper = new Beeper();
        Thread hBeeperThread = null;
        ManualResetEvent beepEvent = new ManualResetEvent(false);

        public BeeperLocator()
        {
            // Start Beeper thread
            beeperRate = 0;
            beeperRunning = true;
            hBeeperThread = new Thread(new ThreadStart(delegate()
            {
                while (beeperRunning)
                {
                    beepEvent.Reset();
                    if (beeperStopped)
                    {
                        // State: Stopped
                        beepEvent.WaitOne();
                    }
                    else
                    {
                        // State: Locating
                        if (beeperRate > 0)
                        {
                            // Tag in the range
                            beeper.Hz = 1568;
                            if (beeperRate >= 100)
                            {
                                beeper.Beep(100, false);
                            }
                            else
                            {
                                beeper.Beep(50, false);
                                beepEvent.WaitOne(250 - beeperRate, false);
                            }
                        }
                        else
                        {
                            // No tag in the range
                            beeper.Hz = 1046;
                            beeper.Beep(50, false);
                            beepEvent.WaitOne(2000, false);
                        }
                    }
                }
                hBeeperThread = null;
            }));
            hBeeperThread.IsBackground = true;
            hBeeperThread.Priority = ThreadPriority.BelowNormal;
            hBeeperThread.Start();
        }

        public void Dispose()
        {
            // Stop and Exit Beeper thread
            beeperRunning = false;
            beepEvent.Set();
        }

        public void Beep(int level, bool force)
        {
            beeperRate = (level * 100) / maxLevel;
            if (force)
                beepEvent.Set();
        }

        public int MaxLevel
        {
            get { return maxLevel; }
            set { maxLevel = value; }
        }

        public void Start()
        {
            beeperStopped = false;
            beepEvent.Set();
        }

        public void Stop()
        {
            beeperStopped = true;
            beepEvent.Set();
        }
    }
}
